src/Features/Core/Portable/Completion/SharedSyntaxContextsWithSpeculativeModel.cs

```diff
@@ -5,47 +5,43 @@
-        private readonly Lazy<ImmutableArray<DocumentId>> _lazyRelatedDocumentIds;
+    private readonly ConcurrentDictionary<Document, Task<SyntaxContext>> _cache;
+    private readonly Lazy<ImmutableArray<DocumentId>> _lazyRelatedDocumentIds;
 
-        public SharedSyntaxContextsWithSpeculativeModel(Document document, int position)
-        {
-            _document = document;
-            _position = position;
-            _cache = new();
-            _lazyRelatedDocumentIds = new(_document.GetLinkedDocumentIds, isThreadSafe: true);
-        }
+    public SharedSyntaxContextsWithSpeculativeModel(Document document, int position)
+    {
+        _document = document;
+        _position = position;
+        _cache = new();
+        _lazyRelatedDocumentIds = new(_document.GetLinkedDocumentIds, isThreadSafe: true);
+    }
 
-        public Task<SyntaxContext> GetSyntaxContextAsync(Document document, CancellationToken cancellationToken)
+    public Task<SyntaxContext> GetSyntaxContextAsync(Document document, CancellationToken cancellationToken)
+    {
+        if (!_cache.TryGetValue(document, out var contextTask))
         {
-            if (!_cache.TryGetValue(document, out var lazyContext))
-            {
-                if (_document.Id != document.Id && !_lazyRelatedDocumentIds.Value.Contains(document.Id))
-                    throw new ArgumentException("Don't support getting SyntaxContext for document unrelated to the original document");
-
-                lazyContext = GetLazySyntaxContextWithSpeculativeModel(document, this);
-            }
-
-            return lazyContext.GetValueAsync(cancellationToken);
-
-            // Extract a local function to avoid creating a closure for code path of cache hit.
-            static AsyncLazy<SyntaxContext> GetLazySyntaxContextWithSpeculativeModel(Document document, SharedSyntaxContextsWithSpeculativeModel self)
-            {
-                return self._cache.GetOrAdd(document, d => AsyncLazy.Create(cancellationToken
-                    => Utilities.CreateSyntaxContextWithExistingSpeculativeModelAsync(d, self._position, cancellationToken)));
-            }
+            Debug.Assert(_document.Id == document.Id || _lazyRelatedDocumentIds.Value.Contains(document.Id),
+                message: "Don't use for document unrelated to the original document");
+
+            contextTask = GetLazySyntaxContextWithSpeculativeModelAsync(document, this, cancellationToken);
         }
+
+        return contextTask;
+
+        // Extract a local function to avoid creating a closure for code path of cache hit.
+        static Task<SyntaxContext> GetLazySyntaxContextWithSpeculativeModelAsync(Document document, SharedSyntaxContextsWithSpeculativeModel self, CancellationToken cancellationToken)
+            => self._cache.GetOrAdd(document, d => Utilities.CreateSyntaxContextWithExistingSpeculativeModelAsync(d, self._position, cancellationToken));
```

Comment:
this is not safe.  you're capturing the cancellation token into a task which is then cached.  that's why we have AsyncLazy, so that we cache the *computation* but teh cancellation is per-request on that computation.